### PR TITLE
many: add snap-update-ns support for ensure-dir mounts

### PR DESCRIPTION
**Pull request by ernestl**
`snap-update-ns` should create the missing parent directories based on the `ensure-dir` entries generated by `snapd` within the file `snap.<snap-name>.user-fstab`

This PR is `part 3` of a larger feature to automatically create missing personal-files interface parent directories.

There are 3 parts in total:
- Part 1: `pass env real-home and uid to snap-update-ns`              | [pull/13244](https://github.com/snapcore/snapd/pull/13244)
- Part 2: `many: ensure-dir mounts for personal-files missing dirs` | [pull/13260](https://github.com/snapcore/snapd/pull/13260)
- Part 3: `add snap-update-ns support for ensure-dir mounts`        | [pull/13342](https://github.com/snapcore/snapd/pull/13342)

Spread test for all 3 parts: https://github.com/snapcore/snapd/pull/13404

Additional information:
[Relevant section in spec](https://docs.google.com/document/d/1yRqiFcdAIUjwakIQdijtW61vKojcJN2vTjrcwJvLAdo/edit#heading=h.99lwc7f317vt)
Jira [SNAPDENG-8233](https://warthogs.atlassian.net/browse/SNAPDENG-8233)

[SNAPDENG-8233]: https://warthogs.atlassian.net/browse/SNAPDENG-8233?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: strutil/pathiter.go [hunk 1]
```diff
@@ -64,13 +64,13 @@ func (iter *PathIterator) Path() string {
 }
 
 // CurrentName returns the name of the current path element.
-// The return value may end with '/'. Use CleanName to avoid that.
+// The return value may end with '/'. Use CurrentNameNoSlash to avoid that.
```
Comment (pedronis) [inline strutil/pathiter.go:67]: PathInterator could use definitely some more renamings/cleanups but is fine for here I suppose

File: cmd/snap-update-ns/utils.go [hunk 2]
```diff
@@ -339,6 +339,148 @@ func MkSymlink(dirFd int, dirName string, name string, oldname string, rs *Restr
 	return nil
 }
 
+// FileInWayError indicates that a directory cannot be created because
+// a file with the same name exists. This error is intended to be used
+// with MkdirAllWithin.
+type FileInWayError struct {
+	File string
```
Comment (pedronis) [inline cmd/snap-update-ns/utils.go:346]: usually File are os.File  so probably should have a different name like Path or Filename

File: cmd/snap-update-ns/utils.go [hunk 3]
```diff
@@ -339,6 +339,148 @@ func MkSymlink(dirFd int, dirName string, name string, oldname string, rs *Restr
...
+	// Advance the iterator to the first missing directory
+	iter, err := strutil.NewPathIterator(path)
+	if err != nil {
+		return fmt.Errorf("cannot iterate over path %q: %v", path, err)
+	}
+	for iter.Next() {
+		if iter.CurrentPathNoSlash() == parent {
+			break
+		}
+	}
```
Comment (pedronis) [inline cmd/snap-update-ns/utils.go:441]: shouldn't this come before the // Advance the iterator to the first missing directory comment?

File: cmd/snap-update-ns/user.go [hunk 4]
```diff
@@ -33,24 +35,51 @@ type UserProfileUpdateContext struct {
...
-func NewUserProfileUpdateContext(instanceName string, fromSnapConfine bool, uid int) *UserProfileUpdateContext {
+func NewUserProfileUpdateContext(instanceName string, fromSnapConfine bool, uid int) (*UserProfileUpdateContext, error) {
+	realHome, err := snapEnvRealHome()
+	if err != nil {
+		return nil, fmt.Errorf("cannot retrieve home directory: %v", err)
+	}
+	// See bootstrap.c function switch_to_privileged_user(). When snap-update-ns is invoked for
+	// user mounts, the effective uid and gid is changed to the calling user and supplementary
+	// groups dropped, while retaining capability SYS_ADMIN. Having the effective uid and gid
+	// changed to the calling user is a prerequisite for isPlausibleHome to function as intended.
```
Comment (pedronis) [inline cmd/snap-update-ns/user.go:69]: as discussed we need to clarify whether the check here just helps failing early, or if has real security implications?

File: cmd/snap-update-ns/snapenv.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+// snapEnv returns the environment variable SNAP_REAL_HOME provided by the calling process.
+func snapEnvRealHome() (string, error) {
+	env, err := osutilOSEnvironmentUnescapeUnsafe(snapenv.PreservedUnsafePrefix)
+	if err != nil {
+		return "", err
+	}
+
+	const snapRealHome = "SNAP_REAL_HOME"
+	realHome, ok := env[snapRealHome]
```
Comment (pedronis) [inline cmd/snap-update-ns/snapenv.go:39]: as I asked in the previous PR it's unclear if this could just use os.Getenv instead?

**Review by pedronis — COMMENTED** [review]
did a pass

File: cmd/snap-update-ns/utils.go (hunk 2 above)
Comment (ernestl) [reply cmd/snap-update-ns/utils.go]: Fixed

File: cmd/snap-update-ns/snapenv.go (hunk 5 above)
Comment (ernestl) [reply cmd/snap-update-ns/snapenv.go]: Simplified by using os.Getenv() which is sufficient. Removed snapenv.go and added in user.go.

File: cmd/snap-update-ns/change.go [hunk 6]
```diff
@@ -121,6 +120,23 @@ func (c *Change) createPath(path string, pokeHoles bool, as *Assumptions) ([]*Ch
 		err = MkfileAll(path, mode, uid, gid, rs)
 	case "symlink":
 		err = MksymlinkAll(path, mode, uid, gid, c.Entry.XSnapdSymlink(), rs)
+	case "ensure-dir":
```
Comment (ZeyadYasser) [inline cmd/snap-update-ns/change.go:123]: This needs some unit tests.

File: cmd/snap-update-ns/change.go [hunk 7]
```diff
@@ -192,6 +215,13 @@ func (c *Change) ensureSource(as *Assumptions) ([]*Change, error) {
 	}
 
 	kind := c.Entry.XSnapdKind()
+
+	// Source is not relevant to ensure-dir mounts that are intended for
+	// creating missing directories based on the mount target.
+	if kind == "ensure-dir" {
+		return nil, nil
```
Comment (ZeyadYasser) [inline cmd/snap-update-ns/change.go:222]: This needs a unit test.

**Review by ZeyadYasser — COMMENTED** [review]
Great work, Thank you! some small nitpicks.

File: strutil/pathiter.go (hunk 1 above)
Comment (ernestl) [reply strutil/pathiter.go:67]: Agreed. Created maintenance card: https://warthogs.atlassian.net/browse/SNAPDENG-10719

File: cmd/snap-update-ns/user.go (hunk 4 above)
Comment (ernestl) [reply cmd/snap-update-ns/user.go:70]: This was discussed during RIGA meeting with Samuele/Alex.
1. --usermount/user profiles are only actioned for non-root users
2. At this point all capabilities, except for SYS_ADMIN is dropped, and user is changed irreversibly to the real user and real user's supplementary groups are dropped.
3. On systems where apparmor is available, the "owner" rules that also limits access to the actual user home.

In summary,  isPlausibleHome, is only meant for best effort at failing early. If somehow this checkpoint is passed where not intended, it is still not possible to create directories in any location where the DAC access (and apparmor where when available) does not allow the real user permissions to do so.

File: cmd/snap-update-ns/utils.go (hunk 3 above)
Comment (ernestl) [reply cmd/snap-update-ns/utils.go:445]: Improved. There are now 2 comments.

File: cmd/snap-update-ns/utils.go [hunk 8]
```diff
@@ -339,6 +339,148 @@ func MkSymlink(dirFd int, dirName string, name string, oldname string, rs *Restr
...
+	for iter.Next() {
+		if iter.CurrentPathNoSlash() == path {
+			// Already confirmed path does not exist
+			break
+		}
+		fi, err = osLstat(iter.CurrentPathNoSlash())
+		if err == nil {
+			if !fi.Mode().IsDir() {
+				return &FileInWayError{FilePath: iter.CurrentPathNoSlash()}
```
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/utils.go:455]: Thanks for highlighting this, there was multiple unit test cases outstanding. Coverage should now be at 100% for all deltas. This comment serves to address all unit test comments.

File: cmd/snap-update-ns/change.go [hunk 9]
```diff
@@ -171,9 +187,16 @@ func (c *Change) ensureTarget(as *Assumptions) ([]*Change, error) {
 			} else {
 				err = fmt.Errorf("cannot create symlink in %q: existing file in the way", path)
 			}
+		case "ensure-dir":
```
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/change.go:190]: https://github.com/snapcore/snapd/pull/13342/files#r1404221657

File: cmd/snap-update-ns/change.go (hunk 6 above)
Comment (ernestl) [reply cmd/snap-update-ns/change.go:123]: https://github.com/snapcore/snapd/pull/13342/files#r1404221657

File: cmd/snap-update-ns/utils.go [hunk 10]
```diff
@@ -339,6 +339,148 @@ func MkSymlink(dirFd int, dirName string, name string, oldname string, rs *Restr
...
+			return fmt.Errorf("cannot inspect path %q: %v", iter.CurrentPath(), err)
+		}
+		break
+	}
+
+	// Create the first missing directory
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
+	fd, err := sysOpen(iter.CurrentBaseNoSlash(), openFlags, 0)
+	if err != nil {
+		return fmt.Errorf("cannot open directory %q: %v", iter.CurrentBaseNoSlash(), err)
```
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/utils.go:470]: https://github.com/snapcore/snapd/pull/13342/files#r1404221657

File: osutil/mountentry.go [hunk 11]
```diff
@@ -94,6 +94,28 @@ func Unescape(path string) string {
...
 
+// ReplaceMountEntryOption replaces the first mount entry option that has
+// the same prefix as the given option with the given option
+func ReplaceMountEntryOption(entry *MountEntry, option string) {
+	if entry == nil {
+		return
+	}
+	kv := strings.SplitN(option, "=", 2)
+	if len(kv) < 2 {
+		return
```
[the remark this answers is not in the document]
Comment (ernestl) [reply osutil/mountentry.go:105]: I understand you point of view, but not fully confident this would be preferred. @pedronis your view on this would be appreciated.

File: cmd/snap-update-ns/expand.go [hunk 12]
```diff
@@ -45,6 +40,11 @@ func expandPrefixVariable(path, variable, value string) string {
...
 }
 
+// xdgRuntimeDir returns the path to XDG_RUNTIME_DIR for a given user ID.
+func xdgRuntimeDir(uid int) string {
+	return fmt.Sprintf("%s/%d", dirs.XdgRuntimeDirBase, uid)
+}
+
 // expandXdgRuntimeDir expands the $XDG_RUNTIME_DIR variable in the given mount profile.
 func expandXdgRuntimeDir(profile *osutil.MountProfile, uid int) {
 	variable := "$XDG_RUNTIME_DIR"
```
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/expand.go]: Fixed

File: cmd/snap-update-ns/utils.go [hunk 13]
```diff
@@ -339,6 +339,148 @@ func MkSymlink(dirFd int, dirName string, name string, oldname string, rs *Restr
...
+		}
+		fi, err = osLstat(iter.CurrentPathNoSlash())
+		if err == nil {
+			if !fi.Mode().IsDir() {
+				return &FileInWayError{FilePath: iter.CurrentPathNoSlash()}
+			}
+			continue
+		}
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("cannot inspect path %q: %v", iter.CurrentPath(), err)
```
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/utils.go]: Fixed

File: strutil/pathiter.go [hunk 14]
```diff
@@ -82,10 +82,18 @@ func (iter *PathIterator) CurrentPath() string {
...
+// CurrentPathNoSlash returns the same value as CurrentPath with the right slash trimmed.
+func (iter *PathIterator) CurrentPathNoSlash() string {
+	if iter.right > 0 && iter.path[iter.right-1:iter.right] == "/" && iter.path[:iter.right] != "/" {
+		return iter.path[:iter.right-1]
+	}
+	return iter.path[:iter.right]
+}
+
+// CurrentBaseNoSlash returns the prefix of the path that was traversed,
 // excluding the current name.  The result never ends in '/' except if
```
[the remark this answers is not in the document]
Comment (ernestl) [reply strutil/pathiter.go]: Fixed

**Review by alexmurray — APPROVED** [review]
LGTM - thanks @ernestl

File: cmd/snap-update-ns/change.go [hunk 15]
```diff
@@ -121,6 +120,23 @@ func (c *Change) createPath(path string, pokeHoles bool, as *Assumptions) ([]*Ch
 		err = MkfileAll(path, mode, uid, gid, rs)
 	case "symlink":
 		err = MksymlinkAll(path, mode, uid, gid, c.Entry.XSnapdSymlink(), rs)
+	case "ensure-dir":
+		uid = sysGetuid()
+		gid = sysGetgid()
+		if uid == 0 {
```
[the remark this answers is not in the document]
Comment (alexmurray) [reply cmd/snap-update-ns/change.go:126]: oh ok - yep makes sense to keep it consistent with snap-confine.

File: cmd/snap-update-ns/user.go [hunk 16]
```diff
@@ -33,24 +36,51 @@ type UserProfileUpdateContext struct {
...
+	realHome := os.Getenv("SNAP_REAL_HOME")
+	if realHome == "" {
+		return nil, fmt.Errorf("cannot retrieve home directory")
+	}
+	// See bootstrap.c function switch_to_privileged_user(). When snap-update-ns is invoked for
+	// user mounts, the effective uid and gid is changed to the calling user and supplementary
+	// groups dropped, while retaining capability SYS_ADMIN. Having the effective uid and gid
+	// changed to the calling user is a prerequisite for isPlausibleHome to function as intended.
+	if err := isPlausibleHome(realHome); err != nil {
+		return nil, fmt.Errorf("cannot use invalid home directory %q: %v", realHome, err)
```
Comment (ernestl) [inline cmd/snap-update-ns/user.go:72]: I am reconsidering failing early, because it is possible for users to exist without home directories specified, which means snap-update-ns might start failing, even for someone who is not using personal-files or this feature.

If we do not fail early, the problem is that when real home is not available because it does not exist or is invalid, then we end up with mount profile entries that will attempt to create dirs '/$HOME'/... which in all probability will just fail (which solves the problem) because /'$HOME' does not expected to exist, however, it seems more appropriate to avoid attempts to perform actions on unexpanded mount entries at all, which fundamentally leaves us with 2 approaches:

1. filter out/remove all unexpanded mount entries from the profile early when the profile is loaded, and indicate this was done with WARNING: ...
2. modify mount entry to indicate that it should not be acted on by any `changePerformImpl` step because of expansion issue, this could be new option `x-snapd.invalid=$HOME unavailable or invalid`. What I prefer about this approach, it would give a consistent experience flagging problematic entries [here at perform time](https://github.com/snapcore/snapd/blob/master/cmd/snap-update-ns/update.go#L85). We can populate the error using the message in option `x-snapd.invalid=...`

@pedronis, your view on this would be appreciated.

File: cmd/snap-update-ns/change.go [hunk 17]
```diff
@@ -121,6 +120,25 @@ func (c *Change) createPath(path string, pokeHoles bool, as *Assumptions) ([]*Ch
...
+			logger.Noticef("WARNING: cannot perform ensure-dir mount for target %q, not supported for the root user", path)
+		} else {
+			// https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
+			// Mode hints cannot be used here because it does not support specifying all
+			// directory paths within a given directory.
+			mode = 0700
+
+			err = MkdirAllWithin(path, c.Entry.XSnapdMustExistDir(), mode, uid, gid, rs)
```
Comment (ernestl) [inline cmd/snap-update-ns/change.go:134]: @ZeyadYasser, thanks for asking about not using restrictions. Your question made me reconsider and I decided to use restrictions. Note that modeHints could not be used, because it does not allow glob style expression that includes child dirs.

File: cmd/snap-update-ns/user.go (hunk 16 above)
Comment (ZeyadYasser) [reply cmd/snap-update-ns/user.go:72]: Great catch, I more inclined towards the `x-snapd.invalid` approach because it be used for other entries later, but in both approaches we need to consider how to act when users create their homedirs afterwards.

- How NeededChanges will handle entries with x-snapd.invalid in case the homedirs exist?
- Will the user have to wait for an operation (e.g. refresh, connection) that triggers mount entries recalculations (NeededChanges), or can we have something detect this on each run?

Comment (pedronis) [reply cmd/snap-update-ns/user.go:72]: I'm not entirely a fan of creating entries that are just marked invalid,  later phases need to sort things etc, but need to ignore them etc etc. 

> How NeededChanges will handle entries with x-snapd.invalid in case the homedirs exist?

if we don't have them but they now appear, this is just more work to do that it should be able to handle

Comment (ernestl) [reply cmd/snap-update-ns/user.go:72]: Elaborating more on the idea in case it may be useful:

A desired mount entry marked as "invalid"/"unexpanded", should not be treated different from any other desired entry while calculating neededChanges. It serves as an instruction that performing the change (in this case ensuring the target) should not be attempted and must fail.

This means it will be treated the same as any other failed needed change, it will NEVER become part of the list of changes that is saved and loaded as "current" changes at the next invocation.

Comment (ernestl) [reply cmd/snap-update-ns/user.go:72]: It turns out that, regardless of the new snap-update-ns early fail, if the calling user home directory does not exist, snap-confine will error (after returning from snap-update-ns) and the snap application will not execute:

(1) snap cmd run warns during snap run stage:
`cmd_run.go:1046: WARNING: cannot create user data directory: cannot create snap home dir: mkdir /home/ernest3: permission denied`
(2) Error at snap-confine stage, because it cannot create the user directory.
`cannot create user data directory: /home/ernest3/snap/test-personal-files/x1: Permission denied`
See: 
- https://github.com/snapcore/snapd/blob/master/cmd/snap-confine/snap-confine.c#L461-L465
- [dependency on $SNAP_USER_DATA](https://github.com/snapcore/snapd/blob/master/cmd/snap-confine/user-support.c#L29) (same prefix than $SNAP_REAL_HOME)

So in conclusion, the new snap-update-ns early fail, will only change where it fails, but not that it fails.

=> We could consider adding a check for the existence of $SNAP_REAL_HOME earlier in snap-confine, before running snap-update-ns for user mount.

Comment (ernestl) [reply cmd/snap-update-ns/user.go]: @perdonis, here is my interpretation of the concept you suggested:
(1) Still [check validity of user home](https://github.com/snapcore/snapd/pull/13342/files#diff-5862fc64285517730e1ae794ad4f4c455c57533d287e3da20d3110ec6b4fa5d2R71-R79) when creating user update context
(2) [Store the home (valid/invalid/empty) and also the error](https://github.com/snapcore/snapd/pull/13342/files#diff-5862fc64285517730e1ae794ad4f4c455c57533d287e3da20d3110ec6b4fa5d2R89-R90) as part of the profile
(3) When loading the desired profile, [pass home and its error to `expandHomeDir`](https://github.com/snapcore/snapd/pull/13342/files#diff-5862fc64285517730e1ae794ad4f4c455c57533d287e3da20d3110ec6b4fa5d2R121-R123). For non-nil error it would produce an error, _only if where was anything to expand_, that only reference the first mount expansion that fails. This will result in `snap-update-ns` exiting with error.
(4) When assumptions is requested, if there was home error, it [will not unrestrict the home](https://github.com/snapcore/snapd/pull/13342/files#diff-5862fc64285517730e1ae794ad4f4c455c57533d287e3da20d3110ec6b4fa5d2R121-R123).

**Review by pedronis — COMMENTED** [review]
thanks, a few remarks

File: cmd/snap-update-ns/expand.go [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+	for i := range profile.Entries {
+		profile.Entries[i].Name, _ = expandPrefixVariable(profile.Entries[i].Name, envVar, value)
+		profile.Entries[i].Dir, _ = expandPrefixVariable(profile.Entries[i].Dir, envVar, value)
+	}
+}
+
+// expandHomeDir expands the $HOME variable in the given mount profile for entries
+// of mount kind "ensure-dir". It returns and error if expansion is required but the
+// given homeError indicates that given home cannot be used for expansion.
+func expandHomeDir(profile *osutil.MountProfile, home string, homeError error) error {
```
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/expand.go:61]: thanks, changed as proposed.

File: cmd/snap-update-ns/change.go [hunk 19]
```diff
@@ -121,6 +120,25 @@ func (c *Change) createPath(path string, pokeHoles bool, as *Assumptions) ([]*Ch
 		err = MkfileAll(path, mode, uid, gid, rs)
 	case "symlink":
 		err = MksymlinkAll(path, mode, uid, gid, c.Entry.XSnapdSymlink(), rs)
+	case "ensure-dir":
+		uid = sysGetuid()
+		gid = sysGetgid()
+		if uid == 0 {
+			logger.Noticef("WARNING: cannot perform ensure-dir mount for target %q, not supported for the root user", path)
```
[the remark this answers is not in the document]
Comment (pedronis) [reply cmd/snap-update-ns/change.go:127]: I'm still slightly confused here because what snap-confine does is not support a user mount namespace for too, but this is disabling ensure-dir for any kind of namespace but if I remember you didn't make ensure-dir user mount namespace only?

File: cmd/snap-update-ns/expand.go [hunk 20]
```diff
@@ -56,27 +56,29 @@ func expandXdgRuntimeDir(profile *osutil.MountProfile, uid int) {
...
 
 // expandHomeDir expands the $HOME variable in the given mount profile for entries
-// of mount kind "ensure-dir". It returns and error if expansion is required but the
-// given homeError indicates that given home cannot be used for expansion.
-func expandHomeDir(profile *osutil.MountProfile, home string, homeError error) error {
+// of mount kind "ensure-dir". It returns and error if expansion is required but home
+// err indicates that path should not be used for expansion.
+func expandHomeDir(profile *osutil.MountProfile, home func() (path string, err error)) error {
 	const envVar = "$HOME"
+	path, err := home()
```
Comment (pedronis) [inline cmd/snap-update-ns/expand.go:63]: homeErr was probably a better name for an error not checked immediately

**Review by pedronis — COMMENTED** [review]
one nitpick, and one further remark/question following the previous one

File: cmd/snap-update-ns/expand.go (hunk 20 above)
Comment (ernestl) [reply cmd/snap-update-ns/expand.go]: fixed

**Review by ZeyadYasser — APPROVED** [review]
Great work, Thank you!

minor typo comments, nothing huge.

File: cmd/snap-update-ns/expand.go [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+	const envVar = "$XDG_RUNTIME_DIR"
+	value := xdgRuntimeDir(uid)
+	for i := range profile.Entries {
+		profile.Entries[i].Name, _ = expandPrefixVariable(profile.Entries[i].Name, envVar, value)
+		profile.Entries[i].Dir, _ = expandPrefixVariable(profile.Entries[i].Dir, envVar, value)
+	}
+}
+
+// expandHomeDir expands the $HOME variable in the given mount profile for entries
+// of mount kind "ensure-dir". It returns and error if expansion is required but home
```
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/expand.go]: Thanks, fixed

File: cmd/snap-update-ns/user.go [hunk 22]
```diff
@@ -37,12 +37,21 @@ type UserProfileUpdateContext struct {
...
 	uid int
 	// home contains the user's real home directory
-	home string
+	home      string
+	homeError error
 }
 
-// isPlausibleHome returns and error if the path is not clean and absolute path to a directory
-// (symlinks not followed) that can be opened for reading by the user.
+// isPlausibleHome returns and error if the path is empty, not clean, not absolute or cannot
```
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/user.go]: Thanks, fixed

File: cmd/snap-update-ns/utils.go [hunk 23]
```diff
@@ -339,6 +342,151 @@ func MkSymlink(dirFd int, dirName string, name string, oldname string, rs *Restr
...
+
+func (e *ParentMissingError) Is(err error) bool {
+	_, ok := err.(*ParentMissingError)
+	return ok
+}
+
+// MkdirAllWithin is the secure variant of os.MkdirAll that creates all the missing directories of the given path within the
+// given existing parent directory.
+//
+// Unlike the regular version this implementation does not follow any symbolic
```
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/utils.go]: Thanks, fixed

File: cmd/snap-update-ns/change.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (ernestl) [reply cmd/snap-update-ns/change.go]: As discussed, this check is not necessary. Removed the check to enable ensure-dir for root/non-root, even though currently it's only used for user-mounts.

File: cmd/snap-update-ns/change.go [hunk 24]
```diff
@@ -191,11 +180,9 @@ func (c *Change) ensureTarget(as *Assumptions) ([]*Change, error) {
 			}
 		case "ensure-dir":
 			if !fi.Mode().IsDir() {
-				// Failure to complete because of a file in the way should not be treated as an error.
-				logger.Noticef("WARNING: cannot perform ensure-dir mount for target %q: existing file in the way", path)
+				err = fmt.Errorf("cannot create ensure-dir target %q: existing file in the way", path)
```
Comment (pedronis) [inline cmd/snap-update-ns/change.go:183]: what's usually the behavior in these situations in other part of this code? this change seems not one we discussed

**Review by pedronis — COMMENTED** [review]
question

File: cmd/snap-update-ns/change.go (hunk 24 above)
Comment (ernestl) [reply cmd/snap-update-ns/change.go:183]: Yes, this change was not part of the last discussion. 
I initially proposed to only issue warnings if `ensire-dir` target is eclipsed by a file, or `must-exist-dir` does not exist.

The reality is, apart form origin `layout` and `overname` and error `IgnoredMissingMount`, [these errors will be issued as a notice anyways](https://github.com/snapcore/snapd/blob/master/cmd/snap-update-ns/update.go#L85).

Now it follows pattern (e.g. [see kind symlink](https://github.com/snapcore/snapd/pull/13342/files#diff-fa7a82a9939947936744df5b4d0d5b74373e4184265ff52a9d56ae22d28a2c9aR179)) to error if the target dir cannot be created because there is a file in the way. This also simplifies mkDirAllWithin to not require specific error definitions anymore.

**Review by pedronis — APPROVED** [review]
thank you

**Review by ZeyadYasser — APPROVED** [review]
lgtm, thank you!
